### PR TITLE
Update to rules_go 0.12.1 and gazelle 0.12.0 and perform related cleanups

### PR DESCRIPTION
**Pull request by ixdy**
**What this PR does / why we need it**: my initial intent was to simply update to rules_go 0.12.1 and gazelle 0.12.0.

A few internal changes / deprecations meant that I finally needed to clean up some technical debt. This also fixes #64122.
I've attempted to keep the steps as separate commits to make it easier to review:

1. Disable gazelle proto rule generation; legacy proto rules are deprecated, and we don't (currently) build protos at build time anyway, instead generating them with `hack/update-generated-protobuf.sh` and then checking them in. We can revisit this in the future if we'd like.
2. Remove the legacy `go_default_library_protos` filegroups using [buildozer](https://github.com/bazelbuild/buildtools/tree/master/buildozer). We don't use these, anyway.
3. Update the rules_go bazel workspace dependency to 0.12.1.
4. Vendor gazelle 0.12.0 and update BUILD files with `hack/update-bazel.sh`. This causes a lot of diffs, because `select()`s are no longer used in `srcs` attributes, external tests are folded into non-external tests, and vendored targets get an `importmap` attribute.
5. Set `gazelle:prefix` on `staging/src/BUILD` to make gazelle treat these correctly(ish). This allows us to remove the sed rewrite hack in `hack/update-bazel.sh`.
6. Explicitly set `# gazelle:importmap_prefix k8s.io/kubernetes/vendor` on `vendor/`, so that all vendored dependencies get the right importmap. gazelle 0.12.0 uses the bazel workspace name + `vendor/` as a prefix, which doesn't work with native go. Newer gazelle will use the go prefix (https://github.com/bazelbuild/bazel-gazelle/pull/207), but it's not released yet. Setting this correctly now also fixes later `BUILD` churn.
7. Re-run `hack/update-bazel.sh`. This causes a bunch of diffs, since anything under `staging/src` now uses the `staging/src/` path instead of `vendor/`. (Both would work for bazel, but gazelle uses the former, since `vendor/` uses symlinks.) Also `importmap`s under `vendor/` are fixed.
8. Reformat a few files (using [buildifier](https://github.com/bazelbuild/buildtools/tree/master/buildifier)) to make later diffs easier to read.
9. Rework the `go_genrule` rules to use the new `go_genrule` from https://github.com/kubernetes/repo-infra/pull/72, which is more bazely, since it uses the rules_go `go_path` rule instead of lots of shell.
10. Remove the deprecated `go_prefix` rule from the root BUILD.bazel file.
11. Set `# gazelle:importmap_prefix k8s.io/kubernetes/vendor` on `staging/src` as well, which ensures that these repos are treated as vendored dependencies. (It's basically the bazel-y way of doing the `vendor/k8s.io` symlinks.)
12. Run `hack/update-bazel.sh` one last time to fix all of the `importmap`s under `staging/src`.

Note re: point 6 above - we're pretty much ignoring the `vendor/k8s.io` symlinks entirely now under bazel. Using the `gazelle:prefix` directive ensures these get mapped into the right go importpath, and the `go_path` rule installs these correctly now too.

**Special notes for your reviewer**: this should not be submitted before https://github.com/kubernetes/repo-infra/pull/72, obviously.

**Release note**:

```release-note
NONE
```

/assign @BenTheElder @fejta @thockin 
cc @cblecker @jayconrod

### CODE REVIEW

File: build/root/WORKSPACE
Comment (ixdy) [inline build/root/WORKSPACE]: TODO(@ixdy): fix before merge

Comment (BenTheElder) [reply build/root/WORKSPACE]: what's going on here?

Comment (ixdy) [reply build/root/WORKSPACE]: this is my fork containing kubernetes/repo-infra#72. once that merges, I'll need to switch this to latest commit from the repo-infra repo.

Comment (ixdy) [reply build/root/WORKSPACE]: this is fixed now

File: pkg/generated/openapi/def.bzl
Comment (BenTheElder) [inline pkg/generated/openapi/def.bzl]: what's with dropping `srcs` here? did we ever need these for this rule?

Comment (ixdy) [reply pkg/generated/openapi/def.bzl]: I don't think we need them for the generator (currently it's only `doc.go`) - they're only needed for the `go_library` itself.

**Comment by BenTheElder** [comment]
@ixdy PR needs rebase 😬

**Comment by ixdy** [comment]
yeah, I expect at least one rebase per day until this gets approved and merged.

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Comment by BenTheElder** [comment]
needs rebase again

**Comment by ixdy** [comment]
/hold cancel

it's not clear to me this will make it into 1.11, though.

**Comment by thockin** [comment]
Godeps LGTM

/approve

**Comment by ixdy** [comment]
/hold

I'm still not entirely sure about the importmap values for `vendor/` and `staging/src/`, so I'm still playing with that a bit.

**Comment by ixdy** [comment]
I've added a few more commits to this PR, based on discussion with @jayconrod and a better understanding of the `importmap` attribute:
> 6. Explicitly set `# gazelle:importmap_prefix k8s.io/kubernetes/vendor` on `vendor/`, so that all vendored dependencies get the right importmap. gazelle 0.12.0 uses the bazel workspace name + `vendor/` as a prefix, which doesn't work with native go. Newer gazelle will the go prefix (https://github.com/bazelbuild/bazel-gazelle/pull/207), but it's not released yet. Setting this correctly now also fixes later `BUILD` churn.

> 11. Set `# gazelle:importmap_prefix k8s.io/kubernetes/vendor` on `staging/src` as well, which ensures that these repos are treated as vendored dependencies. (It's basically the bazel-y way of doing the `vendor/k8s.io` symlinks.)

> 12. Run `hack/update-bazel.sh` one last time to fix all of the `importmap`s under `staging/src`.

Note that this actually fails to build right now due to a bug I discovered in rules_go, https://github.com/bazelbuild/rules_go/issues/1538.

Given that it's code freeze, we can wait on the fix and then incorporate it into the PR, likely by updating our dependency on rules_go once more.

**Comment by ixdy** [comment]
I think this PR should be good to go now. I updated to rules_go 0.12.1, which should have the needed bug-fix for external tests with custom importmaps.

**Comment by ixdy** [comment]
Thanks past-@ixdy for that comment:
https://github.com/kubernetes/kubernetes/blob/ee7e59b75605ffbb52a9164c81f4077d3fac80b7/pkg/version/def.bzl#L20

(the `vendor/` prefix is needed now, since the importmap is actually correct.)

**Comment by ixdy** [comment]
/hold cancel

this should be good to go now. can I get an lgtm before I have to rebase again? :)

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Pull request merged**
